#!/bin/bash
#PBS -N parse_meta_bcac-white_cimba_erneg
#PBS -S /bin/bash
#PBS -l walltime=12:00:00
#PBS -l mem=32gb
#PBS -l nodes=1:ppn=1

#PBS -o logs/${PBS_JOBNAME}.log

### Standard Error and Output to same file
#PBS -j oe
# Sample size for BRCA1 https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5808456/ 18,908 N 9,414 Case
# Sample size for BCAC https://github.com/privefl/paper-ldpred2/blob/master/code/prepare-sumstats.R N 256,123 137,045 Case

module load gcc/6.2.0
module load python/3.5.3

cd $PBS_O_WORKDIR

python3 /gpfs/data/gao-lab/Julian/software/summary-gwas-imputation/src/gwas_parsing.py \
-gwas_file ../output/metal/final_metal_bcac-white_erneg_cimba1.txt \
-output_column_map MarkerName variant_id -output_column_map Allele2 non_effect_allele -output_column_map Allele1 effect_allele \
-output_column_map Effect effect_size -output_column_map StdErr standard_error -output_column_map Freq1 frequency \
-split_column variant_id '_' chromosome position nea ea --chromosome_format \
--insert_value sample_size 319209 --insert_value n_cases 73067 \
-output_order variant_id panel_variant_id chromosome position effect_allele non_effect_allele frequency pvalue zscore effect_size standard_error sample_size n_cases \
-liftover ../input/hg19ToHg38.over.chain.gz \
-snp_reference_metadata ../input/gtex_v8_eur_filtered_maf0.01_monoallelic_variants.txt.gz METADATA \
-output ../output/metaxcan_gwas_parse/meta_analysis_BCAC_CIMBA_erneg_brca.txt.gz
